Explicit unsigned -> uint16_t casts to avoid conversion warnings

tautschnig opened the pull request.

tautschnig committed:
Explicit unsigned -> uint16_t casts to avoid conversion warnings

smowton commented on src/util/unicode.cpp:
```diff
@@ -167,12 +167,12 @@ std::vector<std::string> narrow_argv(int argc, const wchar_t **argv_wide)
-/// \par parameters: A 16-bit integer
-/// \return A 16-bit integer with bytes swapped
-uint16_t do_swap_bytes(uint16_t x)
+/// \par x Integer the lower two bytes of which are to be swapped
```
\param for a single parameter

smowton commented on src/util/unicode.cpp:
```diff
@@ -167,12 +167,12 @@ std::vector<std::string> narrow_argv(int argc, const wchar_t **argv_wide)
-uint16_t do_swap_bytes(uint16_t x)
+/// \par x Integer the lower two bytes of which are to be swapped
+/// \return The lower 16 bits of the integer with those bytes swapped
+unsigned do_swap_bytes(unsigned x)
```
For my money the old code was better

smowton commented on src/util/unicode.cpp:
```diff
@@ -197,10 +197,10 @@ void utf16_append_code(unsigned int code, bool swap_bytes, std::wstring &result)
     // encode the code in UTF16, possibly swapping bytes.
     code=code-0x10000;
     unsigned int i1=((code>>10) & 0x3ff) | 0xD800;
-    unsigned int a1=(swap_bytes)?do_swap_bytes(static_cast<uint16_t>(i1)):i1;
```
How about instead of widening everything to `uint32_t`, narrow `a1` and `a2` to `uint16_t`?

smowton replied:
Nitpick: while you're fixing this function, `unsigned int` should probably be `uint32_t` throughout, since we have an exact width in mind here (a UCS-4 character)

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a5ce56aef0438e080c6299f6857505b23f7993de).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78358864

tautschnig committed:
Move unicode unit test to CATCH and enable it

tautschnig committed:
Make all unicode operations use native endianness

We do not transfer data to another system, thus using architecture-native
endianness is sufficient.

Fixes: #2559

tautschnig committed:
Remove unused macro BUFSIZE

tautschnig commented:
@kroening, @smowton Please take another look, I have now added a fair number of changes to clean up the endianness story.

LAJW commented:
Just an idea: While we're here, we could add a `narrow_cast`, which would check if we're losing precision when casting. While rare, these kinds of bugs, might be particularly difficult to debug: [C++ Core Guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es46-avoid-lossy-narrowing-truncating-arithmetic-conversions).

tautschnig commented:
@LAJW We've already got (two variants of) `narrow` -- would what you are suggesting amount to adding checks in those procedures?

LAJW commented:
@tautschnig Depends. `gsl::narrow` normally works just like `static_cast`, but in addition it checks for situations when the value stored in the larger-size integer cannot fit in the smaller one. If that's the case, it triggers an invariant or throws, depending on configuration of the GSL library. `gsl::narrow_cast` does not, but it's easier to grep (as you also might be `static_cast`ing classes or constness).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a3a5f3a0ff472b4737e67adc76729e8f16aa2530).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78470462
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

tautschnig commented:
@LAJW What you are saying suggests to me (but I may be misunderstanding) that the current `narrow` should be renamed to `narrow_cast` and a second, invariant-checking `narrow` should be introduced? I'm all up for this, but would just ask for guidance.

tautschnig commented:
@peterschrammel @thk123 @allredj Would you be able to advise on TG's state? I am not entirely surprised as some functions have been renamed, but maybe there's more to it. Thanks!

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 9b9aecff5d2c4c730d2f17dc41188538dd38c316).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78505030
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

smowton commented on unit/util/unicode.cpp:
```diff
@@ -6,20 +6,22 @@ Author: Vojtech Forejt, forejtv@diffblue.com
 #include <util/unicode.h>
 
+// the u8 prefix is only available from VS 2015 onwards
+#if !defined(_MSC_VER) || _MSC_VER >= 1900
```
Check with @kroening but I think this is our minimum compiler on MS environments anyhow.

tautschnig replied:
Well, at least AppVeyor's VS seems to be older than that (it does not support the `u8` prefix).

kroening replied:
I'll do a PR that upgrades AppVeyor to 2013. There are no stakeholders that I know of that still care about VS 2013.

smowton reviewed (approved):
LGTM

LAJW commented:
@tautschnig I'm less concerned with names (as long as they're limited to 2 words) and more with functionality. Rather than renaming, I'd convert them into a template - provide an overload for integers, and then use existing `narrow` for strings. So then you end up with `narrow<output_type>(value)` interface with the ability to provide overloads for specific classes, just like with `std::hash`.

I'm tempted to write something up later today, so feel free to merge this PR, and we can start using said casts later.

tautschnig commented:
@LAJW Thanks a lot for your help! This PR is still awaiting approval from a few more anyway, so you've got time :-)

kroening commented on src/util/unicode.h:
```diff
@@ -22,13 +22,12 @@ std::wstring widen(const char *s);
+std::wstring utf8_to_utf16_native_endian(const std::string &in);
+std::string utf16_native_endian_to_java(const wchar_t ch);
+std::string utf16_native_endian_to_java(const std::wstring &in);
 
```
Now that everything is "native endian", does that need to be mentioned in the name of the function?
It's implicit in any function that takes numbers as arguments.
